pcapng/flags.py

```diff
@@ -0,0 +1,231 @@
+
+    def get(self):
+        val = self.get_bits()
+        try:
+            return self.extra[val]
+        except IndexError:
+            return "[invalid value]"
+
+    def set(self, val):
+        if val in self.extra:
+            self.set_bits(self.extra.index(val))
+        elif isinstance(val, int):
+            self.set_bits(val)
+        else:
+            raise TypeError(
+                "Invalid value {val} for {cls}".format(
+                    val=val, cls=self.__class__.__name__
+                )
+            )
+
+
+# Class representing a single flag schema for FlagWord.
+# 'nbits' defaults to 1, and 'extra' defaults to None.
+FlagField = namedtuple(
+    "FlagField", ("name", "ftype", "nbits", "extra"), defaults=(1, None)
+)
+
+
+class FlagWord(object):
+    """\
+    Class to wrap an integer in bitwise flag/field accessors.
+    """
+
+    __slots__ = [
+        "_nbits",
+        "_value",
+        "_schema",
+    ]
+
+    def __init__(self, schema, nbits=32, initial=0):
+        """
+        :param schema:
+            A list of FlagField objects representing the values to be packed
+            into this object, in order from LSB to MSB of the underlying int
+
+        :param nbits:
+            An integer representing the total number of bits used for flags
+
+        :param initial:
+            The initial integer value of the flags field
+        """
+
+        self._nbits = nbits
+        self._value = initial
+        self._schema = OrderedDict()
+
+        tot_bits = sum([item.nbits for item in schema])
+        if tot_bits > nbits:
+            raise TypeError(
+                "Too many fields for {nbits}-bit field "
+                + "(schema defines {tot} bits)".format(nbits=nbits, tot=tot_bits)
+            )
+
+        bitn = 0
+        for item in schema:
+            if not isinstance(item, FlagField):
+                raise TypeError("Schema must be composed of FlagField objects")
+            if not issubclass(item.ftype, FlagBase):
+                raise TypeError("Expected FlagBase, got {}".format(item.ftype))
+            self._schema[item.name] = item.ftype(self, bitn, item.nbits, item.extra)
+            bitn += item.nbits
+
+    def __int__(self):
+        return self._value
+
+    def __repr__(self):
+        rv = "<{0} (value={1})".format(self.__class__.__name__, self._value)
+        for k, v in self._schema.items():
+            rv += " {0}={1}".format(k, v.get())
+        return rv + ">"
+
+    def __getattr__(self, name):
+        try:
+            v = self._schema[name]
+        except KeyError:
+            raise AttributeError(name)
+        return v.get()
+
+    def __setattr__(self, name, val):
+        try:
+            return object.__setattr__(self, name, val)
+        except AttributeError:
+            pass
+        try:
+            v = self._schema[name]
+        except KeyError:
+            raise AttributeError(name)
+        return v.set(val)
+
+
+if __name__ == "__main__":
```

Comment:
This sounds like it should be on a separate file, to be honest.

Reply:
This was a test aid during development, and I left it in as an example use case. I guess I can remove it now, since the `EPBFlags` implementation is an example that's actually in use.